weave/trace/serialize.py

```diff
@@ -110,3 +99,65 @@ def from_json(obj: Any, project_id: str, server: TraceServerInterface) -> Any:
         return parse_uri(obj)
 
     return obj
+
+
+# _[to/from]_json_custom_weave_type are used to serialize and deserialize
```

Comment:
Reviewers start here! I pulled out the to/from json stuff into common functions and short circuit using the cache. All the code changes are in the new data structures